extensions/src/main/java/org/pentaho/platform/spring/ApplicationEventDispatcher.java

```diff
@@ -0,0 +1,140 @@
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+
+public class ApplicationEventDispatcher extends AbstractApplicationEventMulticaster {
+  public final static String ASYNC_ANNOTATION_QUALIFIER = "dispatcher";
+
+  private ApplicationEventMulticaster asyncMulticaster;
+  private ApplicationEventMulticaster defaultMulticaster;
+
+  private BeanFactory beanFactory;
+
+  @Override
+  public void addApplicationListener( final ApplicationListener listener ) {
+    if ( isAsync( listener ) ) {
+      asyncMulticaster.addApplicationListener( listener );
+    } else {
+      defaultMulticaster.addApplicationListener( listener );
+    }
+  }
+
+  @Override
+  public void removeApplicationListener( final ApplicationListener listener ) {
+    asyncMulticaster.removeApplicationListener( listener );
+    defaultMulticaster.removeApplicationListener( listener );
+  }
+
+  @Override
+  public void removeApplicationListenerBean( final String listenerBeanName ) {
+    asyncMulticaster.removeApplicationListenerBean( listenerBeanName );
+    defaultMulticaster.removeApplicationListenerBean( listenerBeanName );
+  }
+
+  @Override
+  public void removeAllListeners() {
+    defaultMulticaster.removeAllListeners();
+    asyncMulticaster.removeAllListeners();
+  }
+
+  public void multicastEvent( final ApplicationEvent event ) {
+    defaultMulticaster.multicastEvent( event );
+    asyncMulticaster.multicastEvent( event );
+  }
+
+  public void multicastEvent(final ApplicationEvent event, final ResolvableType type ) {
+    defaultMulticaster.multicastEvent( event, type );
+    asyncMulticaster.multicastEvent( event, type );
+  }
+
+  public void setAsyncMulticaster( final ApplicationEventMulticaster asyncMulticaster ) {
+    this.asyncMulticaster = asyncMulticaster;
+  }
+
+  public void setDefaultMulticaster( final ApplicationEventMulticaster defaultMulticaster ) {
+    this.defaultMulticaster = defaultMulticaster;
+  }
+
+  @Override
+  public void setBeanFactory( final BeanFactory beanFactory ) {
+    super.setBeanFactory( beanFactory );
+
+    if ( defaultMulticaster instanceof AbstractApplicationEventMulticaster ) {
+      ( (AbstractApplicationEventMulticaster) defaultMulticaster ).setBeanFactory( beanFactory );
+    }
+
+    if ( asyncMulticaster instanceof AbstractApplicationEventMulticaster ) {
+      ( (AbstractApplicationEventMulticaster) asyncMulticaster ).setBeanFactory( beanFactory );
+    }
+
+    this.beanFactory = beanFactory;
+  }
+
+  @Override
+  public void addApplicationListenerBean( final String listenerBeanName ) {
+    final ApplicationListener listener = beanFactory.getBean( listenerBeanName, ApplicationListener.class );
+    if ( isAsync( listener ) ) {
+      asyncMulticaster.addApplicationListenerBean( listenerBeanName );
+    } else {
+      defaultMulticaster.addApplicationListenerBean( listenerBeanName );
+    }
+  }
+
+  private boolean isAsync( final ApplicationListener listener ) {
```

Comment:
Also, please note - in fact spring is still processing the annotations, the "only added thing" that the "disptacher" does, it selects a proper multicaster that will do the right thing - otherwise it is the multicaster that processes the annotation. So this dispacher also reflects uses the annotations to properly dispatch, and is meant to be a wrapper , and should be removed as soon as we have this capability working by spring itself...(??)